### DISCUSSION TITLE
External event loop

### DISCUSSION THREAD

**Discussion by jcorporation** [cesanta/mongoose#2555] [Ask questions, get answers] [answered]
Is there a recommended way to integrate mongoose in an external event loop or to poll some custom fd's in the poll loop of mongoose?

I have to threads:

- The first threads runs a loop in which I call mg_mgr_poll and another poll for some linux specific eventfd's.
- The second threads notifies the first thread through the eventfd.
- The first thread broadcasts some messages to all connected websocket clients on eventfd POLLIN event.

I read about mg_wakeup() in the documentation, but it seems it is tied to a mongoose connection and can only be used in a request/reply scenario. Do I miss anything?

**Comment by scaprile** [comment]
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

The only exception to that is mg_wakeup(), usage example here: https://mongoose.ws/tutorials/multi-threaded/ . Yes, our assumed usage model is request/reply.
I think it will fit your scenario perfectly without resorting to OS signals.
You can also use https://mongoose.ws/documentation/#queue, but there is no way to wake up the manager with that.

**Reply by jcorporation** [reply]
Thanks for your fast reply. How can the broadcast scenario handled with mg_wakeup()? In this scenario there is no distinct mongoose connection available.

**Reply by scaprile** [reply]
IIUC You broadcast to your clients from the thread that runs the manager, something like this: ?
https://github.com/cesanta/mongoose/blob/9bebe7cf687289335b0d83f1004b6e02e65b4372/examples/timers/main.c#L35-L41
(https://mongoose.ws/documentation/tutorials/timers/)
You would implement something like that snippet here:
https://github.com/cesanta/mongoose/blob/9bebe7cf687289335b0d83f1004b6e02e65b4372/examples/multi-threaded/main.c#L57-L60

But I'm only guessing and rambling, why don't you please elaborate on your usage scenario

**Reply by jcorporation** [reply]
You guessed it correctly. My question is, how can I trigger this with the mg_wakeup() function, if I have no mongoose connection id?

My usage scenario:

One thread that runs mongoose. It provides a websocket endpoint for clients. In this thread there is a function to broadcast messages to all connected websocket clients, like your example above. The clients only connect to the websocket and do not send any messages. The websocket is only used to send messages from server to the clients.

Another thread needs to trigger the above broadcast to all websocket clients. There are no requests from clients to server that can trigger this broadcast.

My current attempt (pseudo code):

Mongoose thread:
```c
web_server_loop(void) {
  while (signal_received == 0) {
    // check for events from other thread
    int cnt = poll(pfds, fds, pfds.len, 50);
    if (cnt > 0) {
       ws_broadcast();
    }
    // mongoose event loop
    mg_mgr_poll(mgr, 50);
  }
}
```

**Reply by scaprile** [reply]
Yes you do have, your listener is also a connection

**Reply by jcorporation** [reply]
Thanks for clarification!

**Comment by scaprile** [comment, marked as answer]
Yes you do have, your listener is also a connection
Just follow https://mongoose.ws/documentation/tutorials/webui-plain/
Check this example, quickly done by mixing from existent examples and tutorials: #2556
